gcloud/connection.py

```diff
@@ -91,14 +95,23 @@ def credentials(self):
     def http(self):
         """A getter for the HTTP transport used in talking to the API.
 
-        :rtype: :class:`httplib2.Http`
-        :returns: A Http object used to transport data.
+        This will return a thread-local :class:`httplib2.Http` instance unless
+        a custom transport has been provided to the `Connection` constructor.
+
+        :rtype: :class:`httplib2.Http` or the custom HTTP transport specifed
+            to the connection constructor.
+        :returns: An `Http` object used to transport data.
         """
-        if self._http is None:
-            self._http = httplib2.Http()
+        if self._http is not None:
+            return self._http
+
+        if not hasattr(self._local, 'http'):
```

Comment:
I suppose when the refcount on the connection drops to zero and  `__del__` is called on the thread-local? This would be the same as the current behavior, just with one http object per thread.